File: LeetCodeProblems/Problems/LeetCode105ConstructBinaryTreeFromPreorderAndInorderTraversal.cs
using LeetCodeProblems.Models;

namespace LeetCodeProblems.Problems;

public class LeetCode105ConstructBinaryTreeFromPreorderAndInorderTraversal : BaseProblemClass
{
    public override void Run()
    {
        int[] preOrder = [3, 9, 20, 15, 7];
        int[] inOrder = [9, 3, 15, 20, 7];
        var root=BuildTree(preOrder, inOrder);
        PrintPreOrder(root);
        Console.WriteLine();
        PrintInOrder(root);
    }

    public TreeNode BuildTree(int[] preorder, int[] inorder)
    {
        var currentPreorderIndex = 0;
        var inorderMap = new Dictionary<int, int>();
        for (var i = 0; i < inorder.Length; i++)
        {
            inorderMap[inorder[i]] = i;
        }

        return ConstructSubTree(preorder, inorderMap, ref currentPreorderIndex, 0, inorder.Length - 1);
    }

    private TreeNode ConstructSubTree(int[] preorder, Dictionary<int, int> inorderMap, ref int currentPreorderIndex, int inorderStartIndex,
        int inorderEndIndex)
    {
        if (inorderStartIndex > inorderEndIndex)
            return null;
        var rootVal = preorder[currentPreorderIndex];
        currentPreorderIndex++;
        var root = new TreeNode(rootVal);
        if (inorderStartIndex == inorderEndIndex)
            return root;
        var inorderIndex = inorderMap[rootVal];
        root.left = ConstructSubTree(preorder, inorderMap, ref currentPreorderIndex, inorderStartIndex, inorderIndex - 1);
        root.right = ConstructSubTree(preorder, inorderMap, ref currentPreorderIndex, inorderIndex + 1, inorderEndIndex);
        return root;
    }

    void PrintPreOrder(TreeNode? root)
    {
        if(root == null)
            return;
        Console.Write($"{root.val} ");
        PrintPreOrder(root.left);
        PrintPreOrder(root.right);

    }

    void PrintInOrder(TreeNode? root)
    {
        if(root == null)
            return;
        PrintInOrder(root.left);
        Console.Write($"{root.val} ");
        PrintInOrder(root.right);
     
    }
}

File: LeetCodeProblemsCpp/LeetCodeProblemsCpp/Problems/LeetCode105ConstructBinaryTreeFromPreorderAndInorderTraversal.cpp
#include "LeetCode105ConstructBinaryTreeFromPreorderAndInorderTraversal.h"

#include <iostream>
#include <map>

void LeetCode105ConstructBinaryTreeFromPreorderAndInorderTraversal::Run()
{
    std::vector<int> preorder{-1};
    std::vector<int> inorder{-1};
    auto root=buildTree(preorder, inorder);
    std::cout<<"Preorder: ";
    TreeNode::PrintPreOrder(root);
    std::cout<<'\n'<<"Inorder: ";
    TreeNode::PrintInOrder(root);
}

TreeNode* LeetCode105ConstructBinaryTreeFromPreorderAndInorderTraversal::buildTree(
    std::vector<int>& preorder, std::vector<int>& inorder)
{
    std::map<int, int> inorderMap;
    int size = static_cast<int>(inorder.size());
    for (auto i = 0; i < size; i++)
        inorderMap[inorder[i]] = i;
    return GenerateSubTree(preorder, inorderMap, 0, 0, size - 1);
}

TreeNode* LeetCode105ConstructBinaryTreeFromPreorderAndInorderTraversal::GenerateSubTree(
    std::vector<int>& preorder, std::map<int, int>& inorderMap,
    int rootPreOrderIndex, int beginInOrderIndex, int endInOrderIndex)
{
    if (beginInOrderIndex > endInOrderIndex)
        return nullptr;
    auto root = new TreeNode(preorder[rootPreOrderIndex]);
    int rootInorderIndex = inorderMap[root->val];
    int leftTreeLength = rootInorderIndex - beginInOrderIndex;

    int leftTreeRootPreOrderIndex = rootPreOrderIndex + 1;

    int rightTreeRootPreOrderIndex = rootPreOrderIndex + leftTreeLength + 1;

    root->left = GenerateSubTree(preorder, inorderMap, leftTreeRootPreOrderIndex, beginInOrderIndex,
                                 rootInorderIndex - 1);
    root->right = GenerateSubTree(preorder, inorderMap, rightTreeRootPreOrderIndex, rootInorderIndex + 1,
                                  endInOrderIndex);
    return root;
}
